SAML2 SSO Integration Support

govind4u opened the pull request:
**Description**
This pull request introduces the implementation of SAML2-based Single Sign-On (SSO) authentication into the skills-service application. It enables users to authenticate via an external SAML2 Identity Provider (IdP), enhancing the security and convenience of user authentication.

Key Features:

**1. SAML2 Authentication:**

      -   Integrated Spring Security's SAML2 authentication mechanism.
      -   Configured SAML2 login flow that supports a single identity provider.    

**2. SAML2 Configuration:**

    -  Added SAML2SecurityConfiguration class annotated with @Conditional(SecurityMode.SAML2Auth) to handle SAML2 security-specific configurations.
    -  Relying Party Registration and security filters are defined using RelyingPartyRegistrationRepository and PortalWebSecurityHelper.

**3. Custom User Details Service:**
    -   Implemented CustomSAML2UserDetailsService which loads user details via UserAuthService.
    -   Mapping of SAML attributes to user details and role assignments.

**4. Utilities:**

    -  Added SAML2Utils to manage authentication conversion, providing methods to transform Saml2Authentication to Authentication and UserInfo.
    -  Automatic user creation or updates upon successful SAML login.

**5. Controller Updates:**

    -  Enhanced UserController endpoints for user existence checks and root privilege granting based on SAML2 authentication using @Conditional(SecurityMode.SAML2Auth).

**6. Test Coverage:**

    -  Added new test cases using Spock framework for the SAML2-related components, including security configuration and utility methods.

**Other Changes:**

-  Updated project dependencies in pom.xml to support SAML2 login.

**Testing & Validation:** 

-  Tested with Okta as the SAML2 IdP.
-  Verified authentication flow, user creation, and role assignment are functioning as expected.

   Let me know for any clarifications.  Feel free to make any adjustments needed for the integration

sudo-may commented on build.txt:
Looks like build.txt was accidentally committed. Could you please remove it?

govind4u replied:
Removed

sudo-may commented on dashboard/src/router/UseGlobalNavGuards.js:
`UseGlobalNavGuards.js` has number of console.log statements; Could you please remove them?

govind4u replied:
Removed

sudo-may commented on dashboard/src/router/UseGlobalNavGuards.js:
```diff
@@ -57,25 +57,33 @@ export const useGlobalNavGuards = () => {
     } else if (
-      !isPki() &&
-      to.path?.toLowerCase() === requestAccountPath &&
+      !isPki() && !isSaml2() &&
```
- `.toLowerCase()` was removed; perhaps it would be better to add `toLowerCase()` to `requestAccountPath` to make the comparison case insensitive? 
- && `!isSaml2()` is not needed either here; we plan to remove the `!isPki()` check in the future

govind4u replied:
Refactored as per your comments

sudo-may commented on dashboard/src/router/UseGlobalNavGuards.js:
```diff
@@ -92,10 +100,11 @@ export const useGlobalNavGuards = () => {
-          const landingPageRoute = { name: getLandingPage() }
-          next(landingPageRoute)
-        }
+       if (to.path === '/' && isLoggedIn()) {
```
`isLoggedIn()` was added; suggesting to revert the change

govind4u replied:
Reverted

sudo-may commented on pom.xml:
```diff
@@ -56,6 +57,19 @@
 
+    <repositories>
+
+        <repository>
```
Looks like the addition of the new repository in pom.xml is not necessary as `spring-security-saml2-service-provider` appears to be in maven central. Please remove the newly added repositories tag

govind4u replied:
Spring Security SAML2 relies on OpenSAML libraries which are not hosted on maven central.  We could also use https://build.shibboleth.net/maven/releases/

sudo-may commented on service/src/main/java/skills/auth/saml/CustomSAML2UserDetailsService.groovy:
```diff
@@ -0,0 +1,39 @@
+import org.springframework.stereotype.Component
+import skills.auth.SecurityMode
+
+class CustomSAML2UserDetailsService implements UserDetailsService {
```
`service/src/main/java/skills/auth/saml/CustomSAML2UserDetailsService.groovy` is identical to `skills.auth.form.LocalUserDetailsService`. Can you please remove `CustomSAML2UserDetailsService`?

govind4u replied:
Removed CustomSAML2UserDetailsService

sudo-may commented on service/src/main/java/skills/auth/saml/SAML2SecurityConfiguration.groovy:
```diff
@@ -0,0 +1,111 @@
+        RelyingPartyRegistration registration = RelyingPartyRegistrations.fromMetadataLocation(assertingPartyMetadataLocation)
+                .registrationId(registrationId)
+                .build()
+        return new InMemoryRelyingPartyRegistrationRepository(registration);
```
noticed that `SAML2SecurityConfiguration.groovy` uses `InMemoryRelyingPartyRegistrationRepository`. Will this work in a multi-node installation with multiple skills-service daemons behind a load balancer, or should we consider alternative strategies like storing sessions in Postgres to support distributed instances?

govind4u replied:
It should work as relyingPartyRegistrationRepository bean fetches the asserting party meta data from Identity provider

sudo-may commented on service/src/main/java/skills/auth/saml/UserController.groovy:
```diff
@@ -0,0 +1,47 @@
+@RestController
+@RequestMapping("/")
+@Slf4j
+class UserController {
```
instead of introducing `UserController.groovy` and duplicating the code from `CreateAccountController` please remove `@Conditional(SecurityMode.NonSAML2Auth)` from `CreateAccountController`, after that `UserController.groovy` can be simply removed

govind4u replied:
Refactored as per your suggestions

sudo-may reviewed (changes requested):
Thank you so much for submitting the SAML2 SSO Integration Support Pull Request! We've added some suggestions below as well as inline comments to the Files Changed to help refine the implementation. Thanks again for your hard work and for helping to improve SkillTree!
- Could you please rebase master branch? Recently, we have updated our GitHub Actions Continuous Integration workflows to execute on pull requests. Please ensure that all the tests are passing successfully.
- Please explore implementing service tests for SAML2 support instead of unit tests. This approach will provide more comprehensive testing and ensure that SAML2 functionality doesn't regress in the future. As outlined in the [Contribution Guide](https://skilltreeplatform.dev/contribution/devEnv.html#day-to-day-development), we aim to write black-box integration tests and supplement with unit tests only when necessary.
  - SkillTree's endpoint integration tests, located in the `skills.intTests` package, stand up the entire skills-service application and validate endpoint responses. To test SAML2 support, we can leverage an external dependency like a SAML2 protocol provider.
  - Here's a proposed approach:
    - Use Docker to set up a SAML2-compatible service: utilize an open-source service like Keycloak (Apache License 2.0), which provides a Docker image: https://www.keycloak.org/getting-started/getting-started-docker.
    - Configure the service for local development and testing: can initiate the setup via command line for local development and configure it as a service in GitHub Actions CI.
    - Write integration tests: Create specs that end with IT, which will run in the `maven-failsafe-plugin` and start a separate instance of skills-service/Spring Boot. In these tests, configure SAML-specific Spring properties. There are examples in the repository, such as `skills.intTests.RankingAndProgressViewsDisabledIT`.

govind4u commented:
> Thank you so much for submitting the SAML2 SSO Integration Support Pull Request! We've added some suggestions below as well as inline comments to the Files Changed to help refine the implementation. Thanks again for your hard work and for helping to improve SkillTree!
> 
> * Could you please rebase master branch? Recently, we have updated our GitHub Actions Continuous Integration workflows to execute on pull requests. Please ensure that all the tests are passing successfully.
> * Please explore implementing service tests for SAML2 support instead of unit tests. This approach will provide more comprehensive testing and ensure that SAML2 functionality doesn't regress in the future. As outlined in the [Contribution Guide](https://skilltreeplatform.dev/contribution/devEnv.html#day-to-day-development), we aim to write black-box integration tests and supplement with unit tests only when necessary.
>   
>   * SkillTree's endpoint integration tests, located in the `skills.intTests` package, stand up the entire skills-service application and validate endpoint responses. To test SAML2 support, we can leverage an external dependency like a SAML2 protocol provider.
>   * Here's a proposed approach:
>     
>     * Use Docker to set up a SAML2-compatible service: utilize an open-source service like Keycloak (Apache License 2.0), which provides a Docker image: https://www.keycloak.org/getting-started/getting-started-docker.
>     * Configure the service for local development and testing: can initiate the setup via command line for local development and configure it as a service in GitHub Actions CI.
>     * Write integration tests: Create specs that end with IT, which will run in the `maven-failsafe-plugin` and start a separate instance of skills-service/Spring Boot. In these tests, configure SAML-specific Spring properties. There are examples in the repository, such as `skills.intTests.RankingAndProgressViewsDisabledIT`.

govind4u commented:
> Thank you so much for submitting the SAML2 SSO Integration Support Pull Request! We've added some suggestions below as well as inline comments to the Files Changed to help refine the implementation. Thanks again for your hard work and for helping to improve SkillTree!
> 
> * Could you please rebase master branch? Recently, we have updated our GitHub Actions Continuous Integration workflows to execute on pull requests. Please ensure that all the tests are passing successfully.
> * Please explore implementing service tests for SAML2 support instead of unit tests. This approach will provide more comprehensive testing and ensure that SAML2 functionality doesn't regress in the future. As outlined in the [Contribution Guide](https://skilltreeplatform.dev/contribution/devEnv.html#day-to-day-development), we aim to write black-box integration tests and supplement with unit tests only when necessary.
>   
>   * SkillTree's endpoint integration tests, located in the `skills.intTests` package, stand up the entire skills-service application and validate endpoint responses. To test SAML2 support, we can leverage an external dependency like a SAML2 protocol provider.
>   * Here's a proposed approach:
>     
>     * Use Docker to set up a SAML2-compatible service: utilize an open-source service like Keycloak (Apache License 2.0), which provides a Docker image: https://www.keycloak.org/getting-started/getting-started-docker.
>     * Configure the service for local development and testing: can initiate the setup via command line for local development and configure it as a service in GitHub Actions CI.
>     * Write integration tests: Create specs that end with IT, which will run in the `maven-failsafe-plugin` and start a separate instance of skills-service/Spring Boot. In these tests, configure SAML-specific Spring properties. There are examples in the repository, such as `skills.intTests.RankingAndProgressViewsDisabledIT`.

Added an integration test for SAML2 using testcontainers

govind4u commented:
@sudo-may  I see lot of UI tests failing. Is it an environment issue?  I need to add below statement  @IgnoreIf({env["SPRING_PROFILES_ACTIVE"] == "pki" }) to SAML2 Integration Test

sudo-may commented on service/src/main/java/skills/controller/PublicConfigController.groovy:
```diff
@@ -66,6 +66,9 @@ class PublicConfigController {
     @Value('${skills.config.ui.enablePageVisitReporting:#{false}}')
     Boolean enablePageVisitReporting
 
+    @Value('${spring.security.saml2.registrationId:saml2TestRegId}')
```
Instead of defaulting the test value can you please specify this property in the test itself (`service/src/test/java/skills/intTests/saml2/SAML2IntegrationIT`)?

sudo-may commented on service/src/test/java/skills/intTests/saml2/SAML2IntegrationIT.groovy:
The idea behind `SAML2IntegrationIT` is exactly what we are looking for from the integration tests; can you please migrate what is currently being tested by `service/src/test/java/skills/auth/saml/UserControllerSpec.groovy` to `service/src/test/java/skills/intTests/saml2/SAML2IntegrationIT.groovy`?

govind4u replied:
Migrated

sudo-may commented on service/src/test/java/skills/intTests/saml2/SAML2IntegrationIT.groovy:
```diff
@@ -0,0 +1,70 @@
+
+
+@SpringBootTest(webEnvironment=SpringBootTest.WebEnvironment.RANDOM_PORT, classes = SpringBootApp)
+class SAML2IntegrationIT extends Specification{
```
Please extends `skills.intTests.utils.DefaultIntSpec`

govind4u replied:
Including SAML2 login in DefaultIntSpec would be pretty intrusive and requires lot of changes. That's why didn't extend DefaultIntSpec.  As you suggested used KeyCloak test container for SAML2 setup and testing the integration cases

sudo-may commented on service/src/test/java/skills/intTests/saml2/SAML2IntegrationIT.groovy:
```diff
@@ -0,0 +1,70 @@
+    def setupSpec() {
+        keycloak.start()
+
+        // Set the SAML IDP metadata for the Keycloak instance
```
Please move configuration inside` @SpringBootTest(properties = [])` on line 30; see `skills.intTests.RankingAndProgressViewsDisabledIT` for an example

govind4u replied:
Some of the properties like application port are dynamic so could not place those in properties array

sudo-may commented on service/src/test/java/skills/intTests/saml2/SAML2IntegrationIT.groovy:
```diff
@@ -0,0 +1,70 @@
+    def "test SAML2 login redirects to Keycloak"() {
+
+        when: "A request is made to the /app/userInfo/settings endpoint"
+        ResponseEntity<String> response = restTemplate.getForEntity("http://localhost:${port}/app/userInfo/settings", String.class)
```
instead of using `RestTemplate` directly please use `skills.intTests.utils.SkillsService`; tests utilize `SkillsService` to interact with endpoints; Please take a look at the tests extending `skills.intTests.utils.DefaultIntSpec`, for example `skills.intTests.quiz.QuizDefCreationValidationSpecs`

govind4u replied:
See my previous comment

sudo-may commented:
@govind4u The issue of the Cypress visual regression tests failing is now resolved on the `master` branch. Feel free to rebase `master` branch and those tests should not pass for you as well.

govind4u commented:
@sudo-may  I have made few changes to integration test case.   I see action required status on the PR.  Can you have a quick look at it once ?

sudo-may commented:
Hi @govind4u, unfortunately I haven't had time to do a deep dive into the integration test updates. But looks like `SAML2IntegrationIT` is currently failing. Would you be kind enough to take a look at the failures? Thank you!

govind4u commented:
> Hi @govind4u, unfortunately I haven't had time to do a deep dive into the integration test updates. But looks like `SAML2IntegrationIT` is currently failing. Would you be kind enough to take a look at the failures? Thank you!

Modified the test case a bit.  If it fails again due to environment dependencies, I need to enable screenshots for debugging the root cause on the CI instance. It's working fine my local environment.

sudo-may commented:
@govind4u Debugging CI failures could be tricky. Tests just finished; looks like `SAML2IntegrationIT` failed with `TimeoutException`; 
```
SAML2IntegrationIT.SAML2 login flow with Keycloak integration:169->getSessionCookieFromWebDriver:234 Timeout Failed to log in and find the user settings menu
```
Since this is a backend service test there is no screenshots as far as I know of. Were there any clues in the skills-service log (`service/target/skills-service-tests.log`) for that job? It can be downloaded from GitHub's Artifact section. If needed, you could then further tweak the config to enable further debugging.

govind4u commented:
> @govind4u Debugging CI failures could be tricky. Tests just finished; looks like `SAML2IntegrationIT` failed with `TimeoutException`;
> 
> ```
> SAML2IntegrationIT.SAML2 login flow with Keycloak integration:169->getSessionCookieFromWebDriver:234 Timeout Failed to log in and find the user settings menu
> ```
> 
> Since this is a backend service test there is no screenshots as far as I know of. Were there any clues in the skills-service log (`service/target/skills-service-tests.log`) for that job? It can be downloaded from GitHub's Artifact section. If needed, you could then further tweak the config to enable further debugging.

Bumped up Selenium version to make use of selenium manager for managing webdriver downloads.  Could you re-run the test cases to check it the failure is happening again.

govind4u commented:
@sudo-may :  SAML2 login test case need chrome browser on the VM for integration test case.  Could you please check whether google chrome is installed on the ubuntu VM in azure?  Made updates to keycloak signing certificate.  Is there a possibility to get temporary access to vm to check the root cause of failure ?

sudo-may commented:
Hi @govind4u, Tests are executed in [GitHub Hosted Runners](https://docs.github.com/en/actions/using-github-hosted-runners/using-github-hosted-runners/about-github-hosted-runners) and there is no direct access to these underlying VMs. Configuration for these workflows can be found in `.github/workflows/` directory. The main on-check-in workflow is `.github/workflows/build-and-test.yml` (which appears to be manually disabled on your fork). Chrome is installed on these VMs and is heavily utilized by Cypress tests so we do know it works.

One of the steps in `.github/workflows/build-and-test.yml` prints the version of the Chrome: 
```
Google Chrome 131.0.6778.69 
``` 
The best way to debug tests that are passing locally but failing in CI is to add/configure additional logging and then review the logs. I know that the process can be a bit frustrating unfortunately! I know that it was my suggestion but if the integration test is adding too much complexity then it would make sense to simply revert to the unit tests and remove the introduced dependencies and certificates.

govind4u commented:
@sudo-may enabled debug logs. Could you please re-run the test suite ?

sudo-may commented:
@govind4u done; however, I believe you can enable the `Continuous Integration` workflow on your fork and run the test suite on your fork https://github.com/govind4u/skills-service/actions

govind4u commented:
> @govind4u done; however, I believe you can enable the `Continuous Integration` workflow on your fork and run the test suite on your fork https://github.com/govind4u/skills-service/actions

@sudo-may At last I was able to run the SAML2 integration test case successfully on my fork.   Looking forward to  successful merge to the master.  Let me know if you need further details

govind4u commented:
@sudo-may :  Can you merge my PR?   I see all tests have passed

sudo-may commented:
> @sudo-may : Can you merge my PR? I see all tests have passed

That's wonderful progress on getting the integration test to pass. However since the integration test was not passing we didn't review the changes after the initial feedback. We'll do our best to review the changes in the near future and get back to you and provide any feedback. Thank you again for the PR.

sudo-may commented on service/src/main/java/skills/auth/saml/SAML2SecurityConfiguration.groovy:
```diff
@@ -0,0 +1,178 @@
+    RelyingPartyRegistrationRepository relyingPartyRegistrationRepository() {
+        RelyingPartyRegistration registration = null;
+
+        if(privateKeyLocation!=null && certificateLocation!=null){
```
It appears that `loadPrivateKey` and `loadCertificate` load a key and a cert from classpath and `loadPrivateKey` replaces contents of the actual key; there should not be production code loading key/cert from the classpath; looks like this code was added as a workaround to get `SAML2IntegrationIT` passing and is not needed for the production runtime; please remove `if(privateKeyLocation!=null && certificateLocation!=null){` statement and its scope as well as auto wiring of these properties; `SAML2IntegrationIT` should be configured to utilize main runtime code of 
```
registration = RelyingPartyRegistrations.fromMetadataLocation(assertingPartyMetadataLocation)
                    .registrationId(registrationId)
                    .build()
```

govind4u replied:
loadPrivateKey and loadCertificate methods retrieve key and cert from both classpath and file system.  Added these to support use cases wherein signing of auth requests is mandatory.  Refactored code to check the value of attribute named 'WantAuthnRequestsSigned' and conditionally load the key and cert for signing the saml request

sudo-may commented on service/src/main/java/skills/auth/form/CreateAccountController.groovy:
```diff
@@ -135,11 +135,10 @@ class CreateAccountController {
         userAuthService.grantRoot(request.getUserPrincipal().name)
     }
 
-    @Conditional(SecurityMode.FormAuth)
```
Instead of removing `@Conditional(SecurityMode.FormAuth)` annotation please change to `@Conditional(SecurityMode.FormOrSAML2Auth)`

govind4u replied:
Done

sudo-may commented on service/src/main/java/skills/controller/PublicConfigController.groovy:
```diff
@@ -107,6 +111,11 @@ class PublicConfigController {
         }
+
+        if(authMode.name() == 'SAML2'){
+            String regId = environment.getProperty("spring.security.saml2.registrationId");
```
Please use the `@Value` annotation method to inject spring property rather than using `environment.getProperty`

govind4u replied:
Done

sudo-may commented on service/pom.xml:
```diff
@@ -362,6 +363,52 @@
+        <dependency>
+            <groupId>org.seleniumhq.selenium</groupId>
+            <artifactId>htmlunit3-driver</artifactId>
+            <version>4.26.0</version>
```
`htmlunit3-driver` needs scope of test

govind4u replied:
Done

sudo-may commented on service/pom.xml:
```diff
@@ -671,6 +718,7 @@
                         <exclude>src/main/resources/templates/**</exclude>
                         <exclude>**/license/*.properties</exclude>
                         <exclude>**/src/main/resources/client-version</exclude>
+                        <exclude>**/src/main/resources/client-version</exclude>
```
duplicate line, please remove

govind4u replied:
Done

sudo-may commented on service/src/test/java/skills/intTests/saml2/SAML2IntegrationIT.groovy:
```diff
@@ -0,0 +1,206 @@
+import java.time.Duration
+
+@IgnoreIf({env["SPRING_PROFILES_ACTIVE"] == "pki" })
+@SpringBootTest(webEnvironment=SpringBootTest.WebEnvironment.RANDOM_PORT, classes = SpringBootApp)
```
Please find attached suggested improvements for the integration test, zip file contains `SAML2IntegrationIT.groovy` and  `realm-export.json`
[SAML2IntegrationIT-updates.zip](https://github.com/user-attachments/files/17938224/SAML2IntegrationIT-updates.zip)

Updates to `SAML2IntegrationIT` include: 
∘ simplification achieved using predefined ports for keycloak and skills-service  `webEnvironment=SpringBootTest.WebEnvironment.DEFINED_PORT` and `keycloak.setPortBindings(["7779:8080"])`
∘ after using predefined ports all of the properties can be loaded on the startup which more accurately mimics production environment
∘ `userExists should return true when user exists` and `userExists should return false when user does not exist` simply mock up `userExists` method then the mocked version was utilized; instead assertions of `/userExists/{user}` endpoint were added to `SAML2 login flow with Keycloak integration` test; also asserting of `/app/userInfo` endpoint was added
∘ added db data reset on startup of the integration tests

govind4u replied:
Made the improvements as suggested with static port

sudo-may commented on dashboard/src/components/access/SSOLogout.vue:
```diff
@@ -0,0 +1,67 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+<script setup> 
```
after another look there is no need for `SSOLogout.vue` component at all; logout can simply redirect to  the`/skills-login` route, then enhance `Login.vue` to optionally show the Login button, similar to what ` <Card v-if="oAuthProviders` card is doing; `SSOLogout.vue` and its route path and associated code can then be removed

govind4u replied:
Removed SSOLogout.vue and modified code to redirect back to /skills-login route

govind4u commented:
@sudo-may: Appreciate your review on the PR.  Incorporated the feedback and ran the test suite on my fork.  Everything seems fine.  Could you please have a quick look at the changes and merge the PR ?
